cardano-node/src/Cardano/Tracing/Tracers.hs

```diff
@@ -594,83 +577,122 @@ forgeTracer
-                                                                   in (cnts { bcBlocksForgedNum = nc }, nc)
-                                                         )
-    traceCounter "blocksForgedNum" updatedBlocksForged tr
-  Consensus.TraceNodeCannotForge {} -> do
+notifyBlockForging fStats tr = Tracer $ \case
+  Consensus.TraceStartLeadershipCheck slot -> do
+    void $ mapForgingCurrentThreadStats slot fStats $
+      \fts -> (fts { ftsLeadershipChecksInitiated = ftsLeadershipChecksInitiated fts + 1 },
+               ftsLeadershipChecksInitiated fts + 1)
+  Consensus.TraceNodeCannotForge slot _ -> do
     -- It means that node have tried to forge new block, but because of misconfiguration
     -- (for example, invalid key) it's impossible.
-    updatedNodeCannotForge <- atomicModifyIORef' bcCounters $ \cnts ->
-      let nc = bcNodeCannotForgeNum cnts + 1
-      in (cnts { bcNodeCannotForgeNum = nc }, nc)
-    traceCounter "nodeCannotForge" updatedNodeCannotForge tr
-  -- The rest of the constructors.
-  _ -> pure ()
-
-notifySlotsMissedIfNeeded
-  :: IORef BlockchainCounters
-  -> Trace IO Text
-  -> Tracer IO (Consensus.TraceForgeEvent blk)
-notifySlotsMissedIfNeeded bcCounters tr = Tracer $ \case
-  Consensus.TraceNodeIsLeader {} -> do
-    updatedNodeIsLeaderNum <- atomicModifyIORef' bcCounters (\cnts -> let nc = bcNodeIsLeaderNum cnts + 1
-                                                                      in (cnts { bcNodeIsLeaderNum = nc }, nc)
-                                                            )
-    traceCounter "nodeIsLeaderNum" updatedNodeIsLeaderNum tr
-  Consensus.TraceNodeNotLeader {} -> do
-    -- Not is not a leader again, so now the number of blocks forged by this node
-    -- should be equal to the number of slots when this node was a leader.
-    counters <- readIORef bcCounters
-    let howManyBlocksWereForged = bcBlocksForgedNum counters
-        timesNodeWasALeader = bcNodeIsLeaderNum counters
-        numberOfMissedSlots = timesNodeWasALeader - howManyBlocksWereForged
-    if numberOfMissedSlots > 0
-    then do
-        -- Node was a leader more times than the number of forged blocks,
-        -- it means that some slots were missed.
-      updatesSlotsMissed <- atomicModifyIORef' bcCounters (\cnts -> let nc = bcSlotsMissedNum cnts + numberOfMissedSlots
-                                                                    in (cnts { bcSlotsMissedNum = nc }, nc)
-                                                          )
-      traceCounter "slotsMissedNum" updatesSlotsMissed tr
-    else return ()
-  -- The rest of the constructors.
+    void $ mapForgingCurrentThreadStats slot fStats $
+      \fts -> (fts { ftsNodeCannotForgeNum = ftsNodeCannotForgeNum fts + 1 },
+               ftsNodeCannotForgeNum fts + 1)
+    sum <$> threadStatsProjection fStats ftsNodeCannotForgeNum
+    >>= traceCounter "nodeCannotForge" tr
```

Comment:
This use of `>>=` in this `do`-block is confusing. At first sight, I would guess it's to avoid naming the intermediary result. But a mix of the two often indicates that they're using a different monad, in which case the `>>=` should definitely be indented further.

Reply:
> I would guess it's to avoid naming the intermediary result

That was to underscore similarity of the case branches (as a hint for further generalisation) -- whereas giving name to variables would be obscuring the similarity.

If that's confusing, I'll remove the pointless-ness.